### DISCUSSION TITLE
Release v3.5.0

### DISCUSSION THREAD

**Discussion by LeStarch** [nasa/fprime#2955] [Announcements]
The release v3.5.0 contains a number of improvements. Primarily, the Operating System Abstraction Layer (OSAL) has been refactored to make integration with new operating systems easier. It also ensures that the OSAL selection for each subsystem is independent, and selected per-executable rather than the entire project.

## Breaking Changes

There are a number of breaking changes in this release. Users will need to fix these issues when upgrading from `v3.4.3`.

### Configuration Changes

Configuration has been substantially changed. Users should migrate to the new `FpConfig.h` available in `fprime/config` and adjust settings accordingly.

### General OSAL Changes

Users are encouraged to call `Os::init()` in the main function of their deployment. While not strictly necessary, this initializes OS singletons at a deterministic time.

> `Fw::Logger::log` calls that precede this call will not be output.

```c++
#include <Os/Os.hpp>
...
int main(...) {
    Os::init();
    ...
}
```

Failing to do so will result in singletons self-initializing on first usage resulting in a very small delay on first usage. `Fw::Logger::log` messages will not use the `Os::Console` output until `Os::init()` is called.

### FPP Changes

FPP has introduced new keywords to support integrated state machines!  This means users who chose those names will need to escape them. Commonly, `state` is used and should be escaped as `$state`

```diff
- event SetBlinkingState(state: Fw.On) ...
+ event SetBlinkingState($state: Fw.On) ...
```

### Task Changes

Most components have standardized on `start`, `stop`, and `join` calls to manage internal tasks. To start these tasks users should switch to the new `start`, `stop`, and `join` methods.

```diff
-    comm.startSocketTask(name, true, COMM_PRIORITY, Default::STACK_SIZE);
+    comm.start(name, true, COMM_PRIORITY, Default::STACK_SIZE);
...

-    comm.stopSocketTask();
-    (void)comm.joinSocketTask(nullptr);
+    comm.join();
+   comm.stop();
```

### Fully Qualified Instance Names

Instances in F Prime dictionaries and typologies are now fully-qualified. This means that the instances are prepended with the module names. To restore the global-instance functionality of older deployments, please remove modules definitions from around instances.

```diff
- module Ref {
...

  instance blockDrv: Drv.BlockDriver base id 0x0100 \
    queue size Default.QUEUE_SIZE \
    stack size Default.STACK_SIZE \
    priority 140
...
- }
```

### StringUtils Changes

Users of StringUtils functions should now supply a `FwSizeType` and may no longer use `U32` as arguments.

### Linux GPIO driver

Users of the LinuxGpioDriver now should exepct a return value of `Drv::GpioStatus` from read and write calls.  Additionally,  the open call now expects a chip argument.

```diff
- bool gpio_success = gpioDriver.open(13, Drv::LinuxGpioDriver::GpioDirection::GPIO_OUT);
+ Os::File::Status gpio_success = gpioDriver.open("/dev/gpiochip0", 13, Drv::LinuxGpioDriver::GpioConfiguration::GPIO_OUTPUT);
```

### Time and Interval Changes

Users should now supply a `Fw::TimeInterval(seconds, microseconds)` to calls to `Os::Task::delay`.  `Svc.TimeVal` has been replaced by `Os::RawTime`. 

## Full List of Changes
* Fixed UT by @SMorettini in https://github.com/nasa/fprime/pull/2543
* Issue 2457 by @LeStarch in https://github.com/nasa/fprime/pull/2502
* Address comments from #2485 by @bocchino in https://github.com/nasa/fprime/pull/2542
* Update File.cpp by @LeStarch in https://github.com/nasa/fprime/pull/2555
* Add shadow variable and pedantic warnings by @JohanBertrand in https://github.com/nasa/fprime/pull/2544
* Add DpManager by @bocchino in https://github.com/nasa/fprime/pull/2546
* Update troubleshooting by @thomas-bc in https://github.com/nasa/fprime/pull/2561
* Fixed documentation of U32 (Swapped signed and unsigned int) by @tsha-256 in https://github.com/nasa/fprime/pull/2559
* Resolve string_copy static analysis warnings by @thomas-bc in https://github.com/nasa/fprime/pull/2556
* Add Fw::ObjectName to hold Fw::ObjBase name by @thomas-bc in https://github.com/nasa/fprime/pull/2497
* Remove FPP dependencies on native int types by @bocchino in https://github.com/nasa/fprime/pull/2548
* Force pip to upgrade during setup action by @thomas-bc in https://github.com/nasa/fprime/pull/2565
* Explicity add test dependencies to Os_ut_exe by @LeStarch in https://github.com/nasa/fprime/pull/2569
* Fix data product array record size by @bocchino in https://github.com/nasa/fprime/pull/2568
* Adding GDS development guide by @LeStarch in https://github.com/nasa/fprime/pull/2577
* Update dictionary JSON spec field names and examples by @jwest115 in https://github.com/nasa/fprime/pull/2574
* Fix crosscompiler path in tutorial by @thomas-bc in https://github.com/nasa/fprime/pull/2599
* Fix broken links in INSTALL.md by @thomas-bc in https://github.com/nasa/fprime/pull/2610
* Fixes #2602; bumps required version to 3.16 by @LeStarch in https://github.com/nasa/fprime/pull/2617
* Update `Svc/PolyDb` to use configurable FPP enumeration as index by @timcanham in https://github.com/nasa/fprime/pull/2587
* Disambiguate fpp error messages. Fixes: #2618 by @LeStarch in https://github.com/nasa/fprime/pull/2620
* Add Java requirement to install guide by @LeStarch in https://github.com/nasa/fprime/pull/2621
* Produce error on restricted targets.  Fixes: #2486 by @LeStarch in https://github.com/nasa/fprime/pull/2633
* Fix missing fpp_depend build. Fixes: #2576 by @LeStarch in https://github.com/nasa/fprime/pull/2631
* Removing GroundInterface component. Fixes: #2037 by @LeStarch in https://github.com/nasa/fprime/pull/2632
* Split CMake tests and run with minimum cmake by @LeStarch in https://github.com/nasa/fprime/pull/2619
* Add DpWriter by @bocchino in https://github.com/nasa/fprime/pull/2593
* Fix INI newline processing. Fixes #2630 by @LeStarch in https://github.com/nasa/fprime/pull/2637
* Correcting rate group documentation. Fixes #1719 by @LeStarch in https://github.com/nasa/fprime/pull/2638
* Issue 1604 by @LeStarch in https://github.com/nasa/fprime/pull/2625
* Fixing Documentation Generation fo macOS by @LeStarch in https://github.com/nasa/fprime/pull/2655
* Adding start apps and arguments by @LeStarch in https://github.com/nasa/fprime/pull/2616
* Initial documentation of fprime libraries by @LeStarch in https://github.com/nasa/fprime/pull/2665
* More GDS plugin guide improvements by @LeStarch in https://github.com/nasa/fprime/pull/2670
* Revise types in generated C++ code by @bocchino in https://github.com/nasa/fprime/pull/2668
* Add CODE_OF_CONDUCT.md by @LeStarch in https://github.com/nasa/fprime/pull/2629
* Update JSON Dictionary spec by @thomas-bc in https://github.com/nasa/fprime/pull/2663
* Format cpp and hpp files in Fw/Types by @bocchino in https://github.com/nasa/fprime/pull/2677
* FPP v2.1.0a7 by @bocchino in https://github.com/nasa/fprime/pull/2676
* Revising integration test api. Fixes: #1902 by @LeStarch in https://github.com/nasa/fprime/pull/2652
* trouble-shooting-guide-2500 by @rlcheng in https://github.com/nasa/fprime/pull/2684
* Trouble shooting guide 2500 by @rlcheng in https://github.com/nasa/fprime/pull/2698
* Improve error display on failure by @JohanBertrand in https://github.com/nasa/fprime/pull/2696
* feat(ci): add markdown link checker by @SauravMaheshkar in https://github.com/nasa/fprime/pull/2651
* code clean up for POSIX. by @kevin-f-ortega in https://github.com/nasa/fprime/pull/2700
* Add JSON dictionary generation to CMake system by @thomas-bc in https://github.com/nasa/fprime/pull/2598
* Add Fw::ExternalString and revise string implementations by @bocchino in https://github.com/nasa/fprime/pull/2679
* Data Product Catalog prototype by @timcanham in https://github.com/nasa/fprime/pull/2667
* Fix string records for data products by @bocchino in https://github.com/nasa/fprime/pull/2697
* Add log on TCP/UDP UT failure and fix UDP UT by @JohanBertrand in https://github.com/nasa/fprime/pull/2705
* Fixed conversion warnings on framework tests by @JohanBertrand in https://github.com/nasa/fprime/pull/2606
* Create TickDriver interface by @chownw in https://github.com/nasa/fprime/pull/2708
* Make single-argument string constructors explicit by @bocchino in https://github.com/nasa/fprime/pull/2707
* Fix copying of data product containers by @bocchino in https://github.com/nasa/fprime/pull/2711
* Update of clang tidy checks by @JohanBertrand in https://github.com/nasa/fprime/pull/2710
* Fix Typo by @inferenceus in https://github.com/nasa/fprime/pull/2716
* Renamed Time Interface File and fixed where it's included by @CombustableLem0n in https://github.com/nasa/fprime/pull/2728
* Add contributing tips by @JohanBertrand in https://github.com/nasa/fprime/pull/2704
* fixed typo by @konerzajakub in https://github.com/nasa/fprime/pull/2736
* Add `-Wno-format-truncation` to root project by @thomas-bc in https://github.com/nasa/fprime/pull/2719
* Os task refactor issue 2526 by @LeStarch in https://github.com/nasa/fprime/pull/2672
* Fixing port selection in TCP UTs by @LeStarch in https://github.com/nasa/fprime/pull/2739
* Work/string search by @timcanham in https://github.com/nasa/fprime/pull/2720
* Remove dead link to kitware blog by @LeStarch in https://github.com/nasa/fprime/pull/2746
* Fix CI bash script for macOS by @JohanBertrand in https://github.com/nasa/fprime/pull/2742
* Create Command Interface by @rmzmrnn in https://github.com/nasa/fprime/pull/2753
* Add Events FPP interface by @menaman123 in https://github.com/nasa/fprime/pull/2749
* refactored FW_OBJ_NAME, FW_QUEUE_NAME, and FW_TASK_NAME. by @rldettloff in https://github.com/nasa/fprime/pull/2751
* Update FPP to v2.1.0a11 by @bocchino in https://github.com/nasa/fprime/pull/2745
* Interface from LinuxSpiDriver by @menaman123 in https://github.com/nasa/fprime/pull/2762
* modified:   Drv/TcpClient/TcpClientComponentImpl.cpp by @menaman123 in https://github.com/nasa/fprime/pull/2759
* Added Channel Interface for Telemetry by @menaman123 in https://github.com/nasa/fprime/pull/2757
* Fixed copy-paste typos. Should be UDP and not TCP. by @kevin-f-ortega in https://github.com/nasa/fprime/pull/2778
* feat: Add prepend flag to autocode register macro by @mosa11aei in https://github.com/nasa/fprime/pull/2777
* Update upcoming events by @LeStarch in https://github.com/nasa/fprime/pull/2767
* Add hyperlinks to Upcoming Events by @thomas-bc in https://github.com/nasa/fprime/pull/2787
* Change TCPServer backlog to 1 by @Lex-ari in https://github.com/nasa/fprime/pull/2785
* Add capability for removing sources after an autocoder by @mosa11aei in https://github.com/nasa/fprime/pull/2786
* Diatixis Method Reorginization for Documentation by @GavinKnudsen in https://github.com/nasa/fprime/pull/2717
* Add SmallSat Tutorial to upcoming events by @thomas-bc in https://github.com/nasa/fprime/pull/2789
* Ignore all HTTP links in Markdown link check by @thomas-bc in https://github.com/nasa/fprime/pull/2793
* Rename EventsInterface to EventInterface by @thomas-bc in https://github.com/nasa/fprime/pull/2792
* FPP v2.1.0a12 by @bocchino in https://github.com/nasa/fprime/pull/2795
* Match markdown of Channels with that of Events by @SiboVG in https://github.com/nasa/fprime/pull/2804
* Add Svc.Version component by @Shivaly-Reddy in https://github.com/nasa/fprime/pull/2747
* Add Hub-Pattern Documentation by @Lex-ari in https://github.com/nasa/fprime/pull/2791
* Revise file name string sizes in AssertFatalAdapter events by @bocchino in https://github.com/nasa/fprime/pull/2796
* Make _fprime_packages a folder where all libraries can live by @mosa11aei in https://github.com/nasa/fprime/pull/2812
* Add phase config code to Ref. #2820 by @LeStarch in https://github.com/nasa/fprime/pull/2821
* Delete RPI legacy Topology.cpp and Components.hpp by @LeStarch in https://github.com/nasa/fprime/pull/2822
* docs: Add AC pre-processor pattern to docs by @mosa11aei in https://github.com/nasa/fprime/pull/2808
* docs: Add documentation for F' subtopologies by @mosa11aei in https://github.com/nasa/fprime/pull/2743
* Refactor OS::Mutex in CMake selection by @thomas-bc in https://github.com/nasa/fprime/pull/2790
* FPP v2.1.0 by @bocchino in https://github.com/nasa/fprime/pull/2811
* Implements #2723: Adds Os::Console to replace Os::Log by @LeStarch in https://github.com/nasa/fprime/pull/2831
* Fix double-promotion in fprime codebase by @JohanBertrand in https://github.com/nasa/fprime/pull/2818
* Updated troubleshooting md by @mprather in https://github.com/nasa/fprime/pull/2843
* Update CI to use Json dictionaries by @Lex-ari in https://github.com/nasa/fprime/pull/2806
* Adding `fflush` call to posix console by @LeStarch in https://github.com/nasa/fprime/pull/2857
* Fix Random UT Failure: PosixFile (#2835) by @rlcheng in https://github.com/nasa/fprime/pull/2860
* Update Ref and RPI for compatibility with FPP #349 (fully qualified component instance names) by @jwest115 in https://github.com/nasa/fprime/pull/2823
* Add sequence dispatcher component by @zimri-leisher in https://github.com/nasa/fprime/pull/2731
* Remove versions from SystemResources and updates to Version component by @Shivaly-Reddy in https://github.com/nasa/fprime/pull/2866
* Improvement of Doxycomments in the CCSDS FDP checksum class by @JackNWhite in https://github.com/nasa/fprime/pull/2855
* Fix ComQueue Does Not Deallocate Buffers on Overflow by @DJKessler in https://github.com/nasa/fprime/pull/2853
* Fix random UT failures on PosixFile by @rlcheng in https://github.com/nasa/fprime/pull/2862
* Phase 1 state machines by @garthwatney in https://github.com/nasa/fprime/pull/2829
* Adding FppTest for overflow: drop, assert, hook by @LeStarch in https://github.com/nasa/fprime/pull/2842
* Fix paths in Cross Compiling Tutorial by @kevin-f-ortega in https://github.com/nasa/fprime/pull/2885
* Upgrade GitHub Actions ahead of NodeJS deprecation by @thomas-bc in https://github.com/nasa/fprime/pull/2890
* Removing component init() functions where unneeded by @LeStarch in https://github.com/nasa/fprime/pull/2893
* Fixing string_utils types to FwSizeType by @LeStarch in https://github.com/nasa/fprime/pull/2884
* Add Open Port Request on Sending Side of IpSocket by @csmith608 in https://github.com/nasa/fprime/pull/2683
* Fix Os::Console singleton to be heapless by @LeStarch in https://github.com/nasa/fprime/pull/2896
* Refactor Os::FileSystem and Os::Directory by @thomas-bc in https://github.com/nasa/fprime/pull/2871
* Refactor Os::Queue into CMake selection by @LeStarch in https://github.com/nasa/fprime/pull/2895
* FP-2784: define trace id type by @Shivaly-Reddy in https://github.com/nasa/fprime/pull/2917
* Fix warnings in tcp/udp UT by @JohanBertrand in https://github.com/nasa/fprime/pull/2744
* Set assert size constant by @DJKessler in https://github.com/nasa/fprime/pull/2921
* Issue #2727: Refactor System Resources OSAL Implmentation  by @LeStarch in https://github.com/nasa/fprime/pull/2922
* Update FPP version to v2.2.0 by @bocchino in https://github.com/nasa/fprime/pull/2932
* OSAL clean-up by @LeStarch in https://github.com/nasa/fprime/pull/2933
* Make CI fast again! by @LeStarch in https://github.com/nasa/fprime/pull/2937
* Fixed F' C header files to be C++ compliant. by @bdshenker in https://github.com/nasa/fprime/pull/2936
* Random UT Failure: PosixFile by @rlcheng in https://github.com/nasa/fprime/pull/2939
* Fix issues with DpContainer and ExternalSerializeBuffer by @bocchino in https://github.com/nasa/fprime/pull/2941
* Add Os::RawTime OSAL implementation, refactor Os::IntervalTimer by @thomas-bc in https://github.com/nasa/fprime/pull/2923
* Improved Linux GPIO Driver Using Chardev by @LeStarch in https://github.com/nasa/fprime/pull/2943
* Add OSAL singleton initialization to `Os::init()` by @thomas-bc in https://github.com/nasa/fprime/pull/2947
* Add Os::Queue FPP models by @thomas-bc in https://github.com/nasa/fprime/pull/2948
* Bump Tools and GDS requirements to v3.5.0a1 by @thomas-bc in https://github.com/nasa/fprime/pull/2949
* Fix Jekyll docs build error from curly braces interpretation by @thomas-bc in https://github.com/nasa/fprime/pull/2951
* Update requirements.txt for release v3.5.0 by @thomas-bc in https://github.com/nasa/fprime/pull/2952
* Correcting shutdown/close usage; refactoring TcpServer by @LeStarch in https://github.com/nasa/fprime/pull/2950
* Making Os::Task::delay use the singleton pattern by @LeStarch in https://github.com/nasa/fprime/pull/2954

## New Contributors
* @tsha-256 made their first contribution in https://github.com/nasa/fprime/pull/2559
* @rlcheng made their first contribution in https://github.com/nasa/fprime/pull/2684
* @SauravMaheshkar made their first contribution in https://github.com/nasa/fprime/pull/2651
* @chownw made their first contribution in https://github.com/nasa/fprime/pull/2708
* @inferenceus made their first contribution in https://github.com/nasa/fprime/pull/2716
* @CombustableLem0n made their first contribution in https://github.com/nasa/fprime/pull/2728
* @konerzajakub made their first contribution in https://github.com/nasa/fprime/pull/2736
* @rmzmrnn made their first contribution in https://github.com/nasa/fprime/pull/2753
* @menaman123 made their first contribution in https://github.com/nasa/fprime/pull/2749
* @rldettloff made their first contribution in https://github.com/nasa/fprime/pull/2751
* @mosa11aei made their first contribution in https://github.com/nasa/fprime/pull/2777
* @Lex-ari made their first contribution in https://github.com/nasa/fprime/pull/2785
* @GavinKnudsen made their first contribution in https://github.com/nasa/fprime/pull/2717
* @SiboVG made their first contribution in https://github.com/nasa/fprime/pull/2804
* @Shivaly-Reddy made their first contribution in https://github.com/nasa/fprime/pull/2747
* @mprather made their first contribution in https://github.com/nasa/fprime/pull/2843
* @zimri-leisher made their first contribution in https://github.com/nasa/fprime/pull/2731
* @DJKessler made their first contribution in https://github.com/nasa/fprime/pull/2853
* @garthwatney made their first contribution in https://github.com/nasa/fprime/pull/2829
* @bdshenker made their first contribution in https://github.com/nasa/fprime/pull/2936

**Full Changelog**: https://github.com/nasa/fprime/compare/v3.4.3...v3.5.0

<hr /><em>This discussion was created from the release <a href='https://github.com/nasa/fprime/releases/tag/v3.5.0'>Release v3.5.0</a>.</em>

**Comment by keck-in-space** [comment]
Is there any guidance on how one might upgrade from their current version of F Prime to the new version? We're on 3.4.2 and would like to upgrade to the 3.5.x series.

**Reply by thomas-bc** [reply]
@keck-in-space the way I would go about it is to start by updating the F´ submodule in your project to the desired version. Then you can walk through the release notes to make sure all the "Breaking Changes" are fixed in your project, and hopefully building after that should succeed. Iteratively building the project and looking at the error message + release notes should help through the process.

**Reply by zimri-leisher** [reply]
I just went through this process @keck-in-space , and unfortunately reverted due to schedule difficulty. I made a note of all the breaking changes I encountered in this ticket: #3128 . Subsequently, @LeStarch added that to the release notes on this page: https://github.com/nasa/fprime/releases/tag/v3.5.0

So I would go ahead and update, and then use that page as a guide. One thing which isn't noted there because it was unintentional was the change of all PRM_SET/SAVE cmd names to PARAM_SET/PARAM_SAVE, so your sequences might be broken.

**Reply by zimri-leisher** [reply]
You may also encounter this issue: #3124

**Reply by keck-in-space** [reply]
Thanks for all the help! This is very thorough. With this in mind, I might wait to upgrade until the new telemetry packet definitions from xml to fpp files are ready before attempting to upgrade.

**Reply by zimri-leisher** [reply]
Sounds good! That way you also will get the new autocoded state machines, which are super exciting and useful!

**Reply by keck-in-space** [reply]
I've heard a little about these, but I haven't been able to find any documentation. Is there any yet? Very excited!

**Reply by LeStarch** [reply]
FPP packet support is expected in v3.7.0 (targeting May).

State machines are slated for v3.6.0 (early February).

**Reply by LeStarch** [reply]
Here is the super detailed state machine document: https://nasa.github.io/fpp/fpp-spec.html#Definitions_State-Machine-Definitions

**Reply by bocchino** [reply]
That citation is to the spec, which has all the details for implementers. You may want to consult the User's Guide first: https://nasa.github.io/fpp/fpp-users-guide.html#Defining-State-Machines. It's written more for users.
